include/seastar/core/metrics_api.hh

```diff
@@ -132,31 +134,48 @@ public:
     const labels_type& labels() const {
         return *_labels;
     }
-    internalized_labels_ref internalized_labels() const {
+    label_type internalized_labels() const {
         return _labels;
     }
-    void update_labels(internalized_labels_ref labels) {
+    void update_labels(label_type labels) {
         _labels = labels;
     }
-    sstring full_name() const;
+    sstring full_name() const {
+        return safe_name(_group + "_" + _name);
+    }
 
-    metric_id copy() const {
+    metric_id_t copy() const {
         return *this;
     }
 
-    bool operator<(const metric_id&) const;
-    bool operator==(const metric_id&) const;
+    bool operator<(const metric_id_t& id2) const {
+        return as_tuple() < id2.as_tuple();
+    }
+
+
+    bool operator==(const metric_id_t & id2) const {
+        return as_tuple() == id2.as_tuple();
+    }
 private:
-    metric_id(const metric_id &) = default;
-    metric_id & operator=(const metric_id &) = default;
+
+    metric_id_t(const metric_id_t &) = default;
+    metric_id_t & operator=(const metric_id_t &) = default;
 
     auto as_tuple() const {
         return std::tie(group_name(), instance_id(), name(), labels());
     }
     group_name_type _group;
     metric_name_type _name;
-    internalized_labels_ref _labels;
+    label_type _labels;
 };
+
+using metric_id = metric_id_t<internalized_labels_ref>;
+using foreign_metric_id = metric_id_t<foreign_ptr<internalized_labels_ref>>;
```

Comment:
I don't understand why we'd want a foreign metric_id. Isn't it too low-level? On remote shards we work with large collections of metrics, not individual metrics ids.

Reply:
Not even per internalized label set, one foreign_ptr per the entire registered metric set.